### PR TITLE
ASoC: SOF: prevent sending identical dai configs

### PR DESCRIPTION
**Pull request by dragosht**
If multiple sof dais use the same BE, each one of them
would issue a new ipc dai configuration message to be sent
to the firmware on restore. Check previously restored links
and prevent configuring the BE multiple times.

Signed-off-by: Dragos Tarcatu <dragos_tarcatu@mentor.com>

### CODE REVIEW

**Comment by plbossart** [comment]
I must admit I don't quite get the issue. Can you elaborate on what cases this 'If multiple sof dais use the same BE" statement refers to? What sort of topology are we talking about?

**Comment by ranj063** [comment]
> I must admit I don't quite get the issue. Can you elaborate on what cases this 'If multiple sof dais use the same BE" statement refers to? What sort of topology are we talking about?

@plbossart this could happen even with our topology say for ex: sof-glk-da7219.m4. The BE dai for both the capture and playback over the headset SSP are the same and would result in sending the IPC twice.

File: sound/soc/sof/pm.c [hunk 1]
```diff
@@ -146,6 +146,23 @@ static int sof_restore_pipelines(struct snd_sof_dev *sdev)
 	list_for_each_entry_reverse(dai, &sdev->dai_list, list) {
 		struct sof_ipc_reply reply;
 		struct sof_ipc_dai_config *config = dai->dai_config;
+		struct snd_sof_dai *pdai;
+		int restored = false;
+
+		list_for_each_entry_reverse(pdai, &sdev->dai_list, list) {
```
Comment (ranj063) [inline sound/soc/sof/pm.c:157]: @dragosht the patch looks good. Could you please add a comment here to make it clear that we're preventing sending the config for the same BE DAI again?

Comment (dragosht) [reply sound/soc/sof/pm.c:157]: @ranj063 Sure. Thanks.

**Comment by dragosht** [comment]
> > I must admit I don't quite get the issue. Can you elaborate on what cases this 'If multiple sof dais use the same BE" statement refers to? What sort of topology are we talking about?
> 
> @plbossart this could happen even with our topology say for ex: sof-glk-da7219.m4. The BE dai for both the capture and playback over the headset SSP are the same and would result in sending the IPC twice.

@plbossart  @ranj063 I've tested it using sof-apl-nocodec, but I think pretty much all the DAIs that use both playback/capture on the same SSP port would end up configuring the SSP twice.

For instance, this is what I'm getting in the kernel logs after doing an `aplay -D hw:0,5 test.wav` , if I were to print each restored DAI link in there:

    [   26.324478] sof-audio-pci 0000:00:0e.0: restore DAI link: NoCodec-5
    [   26.324484] sof-audio-pci 0000:00:0e.0: ipc tx: 0x80010000
    [   26.324578] sof-audio-pci 0000:00:0e.0: ipc tx succeeded: 0x80010000
    [   26.324580] sof-audio-pci 0000:00:0e.0: restore DAI link: NoCodec-5
    [   26.324609] sof-audio-pci 0000:00:0e.0: ipc tx: 0x80010000

And the sof logs would also confirm that the SSP was in fact configured twice ...

    0      2          IPC          26020594.583333        64.322914    src/ipc/handler.c:517    ipc: dai 1,5 -> config 
    0      2          DAI          26020598.125000         3.541667        src/lib/dai.c:64     dai_get(), d = 0xbe02f510, sref = 3
    0      2          SSP          26020601.562500         3.437500 ers/intel/cavs/ssp.c:161    ssp_set_config(), config->format = 0x4001
    0      2          SSP          26020608.385417         6.822917 ers/intel/cavs/ssp.c:702    ssp_set_config(), sscr0 = 0xc1c00f7f, sscr1 = 0xd0
    0      2          SSP          26020611.302083         2.916667 ers/intel/cavs/ssp.c:705    ssp_set_config(), sscr2 = 0x00004002, sspsp2 = 0x0
    0      2          SSP          26020614.114583         2.812500 ers/intel/cavs/ssp.c:707    ssp_set_config(), ssrsa = 0x00000003, sstsa = 0x00
    0      2          DAI          26020617.656250         3.541667        src/lib/dai.c:89     dai_put(), dai = 0xbe02f510, sref = 2
    0      2          DAI          26020622.447917         4.791667      src/audio/dai.c:642    config comp 4 pipe 1 dai 5 type 1
    0      2          DMA          26020626.250000         3.802083 src/drivers/dw/dma.c:187    dw_dma_channel_get(): dma 0 request channel 0
    0      2          DAI          26020632.083333         5.833333      src/audio/dai.c:642    config comp 10 pipe 2 dai 5 type 1
    0      2          DMA          26020635.000000         2.916667 src/drivers/dw/dma.c:187    dw_dma_channel_get(): dma 0 request channel 0
    0      2          IPC          26020697.239583        62.239582    src/ipc/handler.c:517    ipc: dai 1,5 -> config 
    0      2          DAI          26020701.093750         3.854167        src/lib/dai.c:64     dai_get(), d = 0xbe02f510, sref = 3
    0      2          SSP          26020704.322917         3.229167 ers/intel/cavs/ssp.c:161    ssp_set_config(), config->format = 0x4001
    0      2          SSP          26020710.208333         5.885417 ers/intel/cavs/ssp.c:702    ssp_set_config(), sscr0 = 0xc1c00f7f, sscr1 = 0xd0
    0      2          SSP          26020713.177083         2.968750 ers/intel/cavs/ssp.c:705    ssp_set_config(), sscr2 = 0x00004002, sspsp2 = 0x0
    0      2          SSP          26020715.885417         2.708333 ers/intel/cavs/ssp.c:707    ssp_set_config(), ssrsa = 0x00000003, sstsa = 0x00
    0      2          DAI          26020719.322917         3.437500        src/lib/dai.c:89     dai_put(), dai = 0xbe02f510, sref = 2
    0      2          DAI          26020723.906250         4.583333      src/audio/dai.c:642    config comp 4 pipe 1 dai 5 type 1

**Comment by plbossart** [comment]
I get the point with the SSP, but that's a specific case where the RX and TX configurations are shared. There are other dais where the Rx and Tx configs are not the same (typically for the 6-pin version of I2s where bit clock/fsync can be different on RX and TX). So in that case you would want the two configs to be sent even if it's the same back-end

**Comment by ranj063** [comment]
> I get the point with the SSP, but that's a specific case where the RX and TX configurations are shared. There are other dais where the Rx and Tx configs are not the same (typically for the 6-pin version of I2s where bit clock/fsync can be different on RX and TX). So in that case you would want the two configs to be sent even if it's the same back-end

@plbossart in that case wouldnt the dai_index be different?

**Comment by plbossart** [comment]
> > I get the point with the SSP, but that's a specific case where the RX and TX configurations are shared. There are other dais where the Rx and Tx configs are not the same (typically for the 6-pin version of I2s where bit clock/fsync can be different on RX and TX). So in that case you would want the two configs to be sent even if it's the same back-end
> 
> @plbossart in that case wouldnt the dai_index be different?

Why would it? On some platforms you have the choice to configure the same interface in 'synchronous mode', with RX and TX using the same config or 'async' mode with RX and TX using different configs. It's still the same hardware, but with 6 pins instead of 4. I am pretty sure our friend @dbaluta will see such cases for the SSI/eSAI support.

**Comment by ranj063** [comment]
> Why would it? On some platforms you have the choice to configure the same interface in 'synchronous mode', with RX and TX using the same config or 'async' mode with RX and TX using different configs. It's still the same hardware, but with 6 pins instead of 4. I am pretty sure our friend @dbaluta will see such cases for the SSI/eSAI support.

@plbossart so in such cases, the dai config would occur during hw_params?

**Comment by plbossart** [comment]
> > Why would it? On some platforms you have the choice to configure the same interface in 'synchronous mode', with RX and TX using the same config or 'async' mode with RX and TX using different configs. It's still the same hardware, but with 6 pins instead of 4. I am pretty sure our friend @dbaluta will see such cases for the SSI/eSAI support.
> 
> @plbossart so in such cases, the dai config would occur during hw_params?

Not sure why? this is a static board-dependent configuration, not something that depends on user selection of sample/formats.
This is different to what @dragosht was working on, which was the ability to store multiple configs from topology and select the right one depending on hw_params.

**Comment by ranj063** [comment]
> > > Why would it? On some platforms you have the choice to configure the same interface in 'synchronous mode', with RX and TX using the same config or 'async' mode with RX and TX using different configs. It's still the same hardware, but with 6 pins instead of 4. I am pretty sure our friend @dbaluta will see such cases for the SSI/eSAI support.
> > 
> > 
> > @plbossart so in such cases, the dai config would occur during hw_params?
> 
> Not sure why? this is a static board-dependent configuration, not something that depends on user selection of sample/formats.
> This is different to what @dragosht was working on, which was the ability to store multiple configs from topology and select the right one depending on hw_params.

@plbossart I am bit confused about what should happen during resume is TX/RX config is different. With the current code, the last config that is sent to the DSP will overwrite the previous one.

**Comment by plbossart** [comment]
> > > > Why would it? On some platforms you have the choice to configure the same interface in 'synchronous mode', with RX and TX using the same config or 'async' mode with RX and TX using different configs. It's still the same hardware, but with 6 pins instead of 4. I am pretty sure our friend @dbaluta will see such cases for the SSI/eSAI support.
> > > 
> > > 
> > > @plbossart so in such cases, the dai config would occur during hw_params?
> > 
> > 
> > Not sure why? this is a static board-dependent configuration, not something that depends on user selection of sample/formats.
> > This is different to what @dragosht was working on, which was the ability to store multiple configs from topology and select the right one depending on hw_params.
> 
> @plbossart I am bit confused about what should happen during resume is TX/RX config is different. With the current code, the last config that is sent to the DSP will overwrite the previous one.

It'd be an interesting case to test. E.g. we could cook a nocodec test where the playback is at 48kHz but the capture at 8kHz only and see what happens. If userspace never uses the two capture and playback paths concurrently it's perfectly valid even with the SSP.

**Comment by dragosht** [comment]
> > > > > Why would it? On some platforms you have the choice to configure the same interface in 'synchronous mode', with RX and TX using the same config or 'async' mode with RX and TX using different configs. It's still the same hardware, but with 6 pins instead of 4. I am pretty sure our friend @dbaluta will see such cases for the SSI/eSAI support.
> > > > 
> > > > 
> > > > @plbossart so in such cases, the dai config would occur during hw_params?
> > > 
> > > 
> > > Not sure why? this is a static board-dependent configuration, not something that depends on user selection of sample/formats.
> > > This is different to what @dragosht was working on, which was the ability to store multiple configs from topology and select the right one depending on hw_params.
> > 
> > 
> > @plbossart I am bit confused about what should happen during resume is TX/RX config is different. With the current code, the last config that is sent to the DSP will overwrite the previous one.
> 
> It'd be an interesting case to test. E.g. we could cook a nocodec test where the playback is at 48kHz but the capture at 8kHz only and see what happens. If userspace never uses the two capture and playback paths concurrently it's perfectly valid even with the SSP.

But is that possible without having 2 separate hw_configs for the SSP to switch between? Is there some way for the playback and capture DAI links of the same SSP to currently end up with different configs?

**Comment by plbossart** [comment]
> But is that possible without having 2 separate hw_configs for the SSP to switch between? Is there some way for the playback and capture DAI links of the same SSP to currently end up with different configs?

It's a bit academic and like I said it would only work when the two paths are not active simultaneously. For other platforms it's definitively possible in full-duplex mode to have RX and TX using separate configs. If the topology and/or kernel don't support this it's a large miss in our SOF design.

**Comment by ranj063** [comment]
> > But is that possible without having 2 separate hw_configs for the SSP to switch between? Is there some way for the playback and capture DAI links of the same SSP to currently end up with different configs?

> It's a bit academic and like I said it would only work when the two paths are not active simultaneously. For other platforms it's definitively possible in full-duplex mode to have RX and TX using separate configs. If the topology and/or kernel don't support this it's a large miss in our SOF design.

@plbossart we also need the same support in the FW too isnt it?

**Comment by plbossart** [comment]
> > > But is that possible without having 2 separate hw_configs for the SSP to switch between? Is there some way for the playback and capture DAI links of the same SSP to currently end up with different configs?
> 
> > It's a bit academic and like I said it would only work when the two paths are not active simultaneously. For other platforms it's definitively possible in full-duplex mode to have RX and TX using separate configs. If the topology and/or kernel don't support this it's a large miss in our SOF design.
> 
> @plbossart we also need the same support in the FW too isnt it?

yep.

**Pull request closed without merging**